docs/migration-guides/v2/lib.md

```diff
@@ -0,0 +1,375 @@
+
+If you’re using the `Crypto.generateRandomKey()` method, you’ll need to change how you call this method. In v1, this method required that you pass a callback. In v2, it communicates its result by returning a promise.
+
+So you need to change code that looks like this:
+
+```javascript
+Ably.Realtime.Crypto.generateRandomKey((err, key) => { … })
+```
+
+into code that makes use of the returned promise, for example by using the `await` keyword on it:
+
+```javascript
+const key = await Ably.Realtime.Crypto.generateRandomKey();
+```
+
+#### Update your usage of `request()` to pass a `version` argument
+
+**Note:** This section is only relevant if you’re using the `Rest.request()` or `Realtime.request()` methods; that is, if you’re using the library to manually make a request to the Ably REST API.
+
+The signature of this method has been changed; it now requires that you pass a `version` argument to specify the major version of the REST API to use. For more information about REST API versioning, see [this section of the REST API reference](https://ably.com/docs/api/rest-api#versioning).
+
+(**TODO this wording "major version" comes from https://github.com/ably/sdk-api-reference/pull/32#pullrequestreview-1408872163. I guess that it means you should take the current sentence 'By default, all requests receive the latest version of the API, which is currently 1.1' from the linked versioning page and infer that the "major version" is 1, and so that's what I should pass to `Rest.request()`. But I thought that we've moved away from using a `major.minor` protocol version to just a single integer? See https://sdk.ably.com/builds/ably/specification/main/features/#CSV2a. Should we update those versioning docs to say that the latest version of the API is 3, and then drop the "major" in the documentation for `Rest.request()`?**)
+
+As an example, given the current code to [get the service time](https://ably.com/docs/api/rest-api#time):
+
+```javascript
+const time = await rest.request('get', '/time');
+```
+
+add an argument to specify the API version:
+
+```javascript
+const time = await rest.request('get', '/time', 3);
+```
+
+#### Update your usage of the result of the `stats()` method
+
+**Note:** This section is only relevant if you’re using the `Rest.stats()` or `Realtime.stats()` method; that is, if you’re using the library to retrieve your application’s usage statistics.
+
+The `Stats` type returned by the `Rest.stats()` and `Realtime.stats()` method has been simplified in v2. Specifically, there is a new property called `entries`, which replaces the following properties:
+
+- `all`
+- `apiRequests`
+- `channels`
+- `connections`
+- `inbound`
+- `outbound`
+- `persisted`
+
+You should migrate your code to use the `entries` property instead of these properties. `entries` is an object all of whose properties have numeric values. Here is an example of how this might look:
```

Comment:
@SimonWoolf I've written this guidance for how to migrate to using the flattened stats API but it feels like a bit of a cop-out; it basically just says "Here's an example of how to migrate one property; beyond that you're own your own. Go fetch some stats, look at the schema they point to, figure out how the changes you need to make".

Can we do any better than this? Are there some clear rules for how the v1 stats API maps to the flattened stats response? For example, it seems that the message-related properties (`all`, `inbound`, `outbound`, `persisted`) now require you to prefix them with `messages`.

Reply:
Ideally this should be in the main docs rather than ably-js specific, but sure, the main changes are:

- the various stats that count messages in different ways (all, inbound, outbound, persisted, processed) are now under `messages` instead of all in the top level
- `connections` is no longer broken down into plain and tls
- instead of top-level `apiRequests` and `tokenRequests`, which doesn't make much sense because token requests are api requests, you now have a top level `apiRequests` that's broken down currently into `tokenRequests` and `other` with an `all` aggregate, with potential to split `other` out into more specific types later
- `push` is completely reorganised in a way that makes more sense

You could also link to [the source of the stats schema](https://github.com/ably/ably-common/blob/main/json-schemas/src/app-stats.json) which is an up to date & accurate description of the new stats struct

Reply:
Thanks! I’ve incorporated what you said above.